File: benchmark/nqueens.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

int64_t abs_int(int64_t x) {
    return (x < 0) ? -x : x;
}

// Check if placing queen at (row, col) conflicts with queen at check_row
int conflicts(int64_t *board, int64_t row, int64_t col, int64_t check_row) {
    int64_t queen_col = board[check_row];
    if (queen_col == col) {
        return 1;
    }
    int64_t row_diff = abs_int(row - check_row);
    int64_t col_diff = abs_int(col - queen_col);
    return row_diff == col_diff;
}

// Check if placing queen at (row, col) is safe
int is_safe_loop(int64_t *board, int64_t row, int64_t col, int64_t check_row) {
    while (check_row < row) {
        if (conflicts(board, row, col, check_row)) {
            return 0;
        }
        check_row++;
    }
    return 1;
}

// Forward declaration
int64_t solve_row(int64_t *board, int64_t row, int64_t n);

// Solve from given row, trying each column
int64_t solve_col(int64_t *board, int64_t row, int64_t n, int64_t col) {
    if (col >= n) {
        return 0;
    }
    
    if (is_safe_loop(board, row, col, 0)) {
        board[row] = col;
        int64_t sub_solutions;
        if (row == (n - 1)) {
            sub_solutions = 1;
        } else {
            sub_solutions = solve_row(board, row + 1, n);
        }
        int64_t rest = solve_col(board, row, n, col + 1);
        return sub_solutions + rest;
    } else {
        return solve_col(board, row, n, col + 1);
    }
}

int64_t solve_row(int64_t *board, int64_t row, int64_t n) {
    return solve_col(board, row, n, 0);
}

int main(void) {
    int64_t n = 12;
    int64_t *board = (int64_t *)malloc(n * sizeof(int64_t));
    
    int64_t solutions = solve_row(board, 0, n);
    printf("Solutions: %lld\n", solutions);
    
    free(board);
    return 0;
}

File: benchmark/nqueens.go
package main

import (
	"fmt"
)

func absInt(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

// Check if placing queen at (row, col) conflicts with queen at check_row
func conflicts(board []int64, row, col, checkRow int64) bool {
	queenCol := board[checkRow]
	if queenCol == col {
		return true
	}
	rowDiff := absInt(row - checkRow)
	colDiff := absInt(col - queenCol)
	return rowDiff == colDiff
}

// Check if placing queen at (row, col) is safe
func isSafeLoop(board []int64, row, col int64) bool {
	for checkRow := int64(0); checkRow < row; checkRow++ {
		if conflicts(board, row, col, checkRow) {
			return false
		}
	}
	return true
}

// Solve from given row, trying each column
func solveCol(board []int64, row, n, col int64) int64 {
	if col >= n {
		return 0
	}

	if isSafeLoop(board, row, col) {
		board[row] = col
		var subSolutions int64
		if row == (n - 1) {
			subSolutions = 1
		} else {
			subSolutions = solveRow(board, row+1, n)
		}
		rest := solveCol(board, row, n, col+1)
		return subSolutions + rest
	}
	return solveCol(board, row, n, col+1)
}

func solveRow(board []int64, row, n int64) int64 {
	return solveCol(board, row, n, 0)
}

func main() {
	n := int64(12)
	board := make([]int64, n)

	solutions := solveRow(board, 0, n)
	fmt.Printf("Solutions for %d-queens: %d\n", n, solutions)
}
